src/trace_ng/log.zig

```diff
@@ -0,0 +1,1048 @@
+        pub fn warnfWithFields(self: *Self, comptime fmt: []const u8, args: anytype, fields: anytype) void {
+            self.logfWithFields(.warn, fmt, args, fields);
+        }
+
+        pub fn info(self: *Self, message: []const u8) void {
+            self.log(.info, message);
+        }
+
+        pub fn infof(self: *Self, comptime fmt: []const u8, args: anytype) void {
+            self.logf(.info, fmt, args);
+        }
+
+        pub fn infoWithFields(self: *Self, message: []const u8, fields: anytype) void {
+            self.logWithFields(.info, message, fields);
+        }
+
+        pub fn infofWithFields(self: *Self, comptime fmt: []const u8, args: anytype, fields: anytype) void {
+            self.logfWithFields(.info, fmt, args, fields);
+        }
+
+        pub fn debug(self: *Self, message: []const u8) void {
+            self.log(.debug, message);
+        }
+
+        pub fn debugf(self: *Self, comptime fmt: []const u8, args: anytype) void {
+            self.logf(.debug, fmt, args);
+        }
+
+        pub fn debugWithFields(self: *Self, message: []const u8, fields: anytype) void {
+            self.logWithFields(.debug, message, fields);
+        }
+
+        pub fn debugfWithFields(self: *Self, comptime fmt: []const u8, args: anytype, fields: anytype) void {
+            self.logfWithFields(.debug, fmt, args, fields);
+        }
+
+        pub fn log(self: *Self, level: Level, message: []const u8) void {
+            switch (self.*) {
+                .noop => {},
+                inline else => |*impl| impl.*.log(level, message),
+            }
+        }
+
+        pub fn logf(self: *Self, level: Level, comptime fmt: []const u8, args: anytype) void {
+            switch (self.*) {
+                .noop => {},
+                inline else => |impl| impl.logf(level, fmt, args),
+            }
+        }
+
+        pub fn logWithFields(self: *Self, level: Level, message: []const u8, fields: anytype) void {
+            switch (self.*) {
+                .noop => {},
+                inline else => |impl| impl.logWithFields(level, message, fields),
+            }
+        }
+
+        pub fn logfWithFields(self: *Self, level: Level, comptime fmt: []const u8, args: anytype, fields: anytype) void {
+            switch (self.*) {
+                .noop => {},
+                inline else => |impl| impl.logfWithFields(level, fmt, args, fields),
+            }
+        }
+    };
+}
+
+pub const Logger = ScoppedLogger(null);
+
+/// An instance of `ScopedLogger` that logs to the standard err.
+pub fn StandardErrLogger(comptime scope: ?[]const u8) type {
+    return struct {
+        const Self = @This();
+        max_level: Level,
+        exit_sig: std.atomic.Value(bool),
+        allocator: Allocator,
+        log_allocator: Allocator,
+        log_allocator_state: *RecycleFBA(.{}),
+        max_buffer: u64,
+        channel: *Channel(logfmt.LogMsg),
+        handle: ?std.Thread,
+
+        pub fn init(config: Config) !*Self {
```

Comment:
Is there a particular reason why this this returns a pointer to self? It could just return `Self` instead of `*Self`, which would simplify the function. Then the calling scope has the choice of how to allocate it. Currently it's forced into allocating the logger with the same allocator that's used for log messages, which may not always be a desirable constraint.

Reply:
> Is there a particular reason why this this returns a pointer to self?

If the `union` holds pointers, ie

```
pub fn ScopedLogger(comptime scope: ?[]const u8) type {
    return union(LogKind) {
       ...
        standard: *StandardErrLogger(scope),
        testing: *TestingLogger(scope),
        ...
}
```

And `ScopedLogger.init(...)` needs to be able to create one of the union variants, then it needs to return a a pointer to self.

With the suggestion here https://github.com/Syndica/sig/pull/249#discussion_r1751166855 might be possible to remove them.